### PR TITLE
Add GET Engine API call

### PR DESCRIPTION
**Pull request by kderusso**
## Closes https://github.com/elastic/enterprise-search-team/issues/3918

Adds a GET Engine API call to the Enterprise Search Engines module. The API pulls the Engine document from the .engine system index. 

To test this: 
1. Enable the feature flag `xpack.ent-search.enabled=true` and ensure all required licensing and authentication is set. 
2. Create an Engine. Using the PUT API (outside this PR) execute a call to upsert the engine. This will automatically create the required system index and alias. Example: 
```
curl --location --request PUT 'http://localhost:9200/_engine/puggles' \
--<credentials>
--header 'Content-Type: application/json' \
--data-raw '{
    "indices": ["test1", "test2"]
}'
```

3. Get the Engine. Example: 
```
curl --location --request GET 'http://localhost:9200/_engine/puggles' \
--<credentials>
--header 'Content-Type: application/json'
```

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java [hunk 1]
```diff
@@ -49,25 +50,42 @@ public class Engine implements Writeable, ToXContentObject {
      * @param name                    The name of the engine.
      * @param indices                 The list of indices targeted by this engine.
      * @param analyticsCollectionName The name of the associated analytics collection.
+     * @param updatedAt               Timestamp this Engine was last updated. Null uses current timestamp.
      */
-    public Engine(String name, String[] indices, @Nullable String analyticsCollectionName) {
+    public Engine(String name, String[] indices, @Nullable String analyticsCollectionName, Long updatedAt) {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:55]: I went back and forth RE: whether to use `TimeValue` vs. `Long` here, and settled on `Long`. Open to feedback here though.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java [hunk 2]
```diff
@@ -163,19 +186,24 @@ public String[] indices() {
...
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         Engine engine = (Engine) o;
         return name.equals(engine.name)
             && Arrays.equals(indices, engine.indices)
-            && Objects.equals(analyticsCollectionName, engine.analyticsCollectionName);
+            && Objects.equals(analyticsCollectionName, engine.analyticsCollectionName)
+            && updatedAt == engine.updatedAt();
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:201]: I wasn't sure if we wanted to included `updatedAt` in `hashCode()` and `equals()`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class GetEngineAction extends ActionType<GetEngineAction.Response> {
+
+    public static final GetEngineAction INSTANCE = new GetEngineAction();
+    public static final String NAME = "cluster:admin/xpack/engines/get";
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:28]: I don't think this is the correct privilege, but I could use some guidance here. I started with `indices:data/read/engines/get` however this leads to errors since the only allowed indices calls that don't specify the indices are scroll and async. Since we don't want to change core internals if we can help it, what would be the best privilege to add here?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:55]: Can we have a single ctr? Can you also add a comment regarding updatedAt, is it in seconds, milliseconds? Maybe we can name it like `updatedAtInMillis` to make it clear? Can we also use a plain `long` here?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:201]: Lets' keep it yes, it's mainly for tests that we have these functions implemented.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java (hunk 3 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:28]: Since we create the alias at Engine creation we can use `indices:admin/engine/get"`. For indices action, the alias/indices in the request are intercepted by the security layer to check that the user has the permission on them. 
To enable this behaviour, the `Request` below should extend `IndicesRequest.Replaceable` and implements `String[] indices()` and `indices(String... indices)`. Since Engine is not an index, we should return the alias name in `indices()` to handle the security check. 
Does that make sense?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java [hunk 4]
```diff
@@ -41,6 +41,7 @@
 public class Engine implements Writeable, ToXContentObject {
     private final String name;
     private final String[] indices;
+    private final long updatedAt;
     private final String analyticsCollectionName;
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:45]: I wonder if we should also add the aliasName? We don't need to do it in this PR though.

File: rest-api-spec/src/main/resources/rest-api-spec/api/engine.get.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      "paths": [
+        {
+          "path": "/_engine/{engine_id}",
+          "methods": [
+            "GET"
+          ],
+          "parts": {
+            "engine_id": {
+              "type": "string",
+              "description": "The name of the engine to be created or updated"
```
Comment (jimczi) [inline rest-api-spec/src/main/resources/rest-api-spec/api/engine.get.json:28]: nit: `The name of the engine`?

**Review by jimczi — COMMENTED** [review]
Thanks @kderusso , the PR looks good. I left a comment regarding the action name and security.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java [hunk 6]
```diff
@@ -77,18 +95,22 @@ public void writeTo(StreamOutput out) throws IOException {
             @SuppressWarnings("unchecked")
             final String[] indices = ((List<String>) params[0]).toArray(String[]::new);
             final String analyticsCollectionName = (String) params[1];
-            return new Engine(engineName, indices, analyticsCollectionName);
+            final long updatedAt = (Long) params[2];
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:98]: This fails the test here:
```
Caused by: java.lang.NullPointerException: Cannot invoke \"java.lang.Long.longValue()\" because \"params[2]\" is null
23:40:27   1> 	at org.elasticsearch.xpack.entsearch.engine.Engine.lambda$static$1(Engine.java:98)
23:40:27   1> 	at org.elasticsearch.xcontent@8.8.0-
```
We need to test if `params[2]` is null first or use a `Long`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 4 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:45]: For now the alias is always using the same prefix - I don't think we need that unless users are able to change it.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportGetEngineAction.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            public void onResponse(Engine engine) {
+                listener.onResponse(new GetEngineAction.Response(engine));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportGetEngineAction.java:42]: I think you can use:

```suggestion
        engineIndexService.getEngine(request.getEngineId(),  listener.map(engine -> new GetEngineAction.Response(engine))));
    }
```

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineActionResponseSerializingTests.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    @Override
+    protected GetEngineAction.Response createTestInstance() {
+        return new GetEngineAction.Response(
+            new Engine(
+                ESTestCase.randomAlphaOfLengthBetween(1, 10),
+                generateRandomStringArray(10, 10, false, false),
+                randomFrom(new String[] { null, randomAlphaOfLengthBetween(1, 10) }),
+                System.currentTimeMillis()
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineActionResponseSerializingTests.java:29]: Nit: Maybe we can use a random, non null long here?

**Review by carlosdelest — COMMENTED** [review]
Looks really good! I'd like to check on the status code from a non-existing engine GET request.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 4 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:45]: Great catch! I was just talking to @TattdCodeMonkey about this earlier in the week. We do need to add the alias name for the front-end (so it's not relying on rules). I was thinking of doing it in a separate PR but if we're changing the serialization we may as well do it now.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportGetEngineAction.java (hunk 7 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportGetEngineAction.java:42]: Ooh thank you, yes, shorter and cleaner is better!

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineActionResponseSerializingTests.java (hunk 8 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineActionResponseSerializingTests.java:29]: Good call, thank you!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 6 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:98]: Thanks for the catch!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java (hunk 3 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:28]: Yes, that does! Thank you for the pointer and explanation!

File: x-pack/plugin/ent-search/qa/rest/roles.yml [hunk 9]
```diff
@@ -2,6 +2,6 @@ entsearch:
   cluster:
     - manage
   indices:
-    - names: '*'
-      privileges:
-        - manage
+    - names: [ "*" ]
+      allow_restricted_indices: true
```
Comment (jimczi) [inline x-pack/plugin/ent-search/qa/rest/roles.yml:6]: Why is this needed (allow_restricted_indices)?

**Review by jimczi — COMMENTED** [review]
Thanks @kderusso , it's getting close!
I left some comments regarding the IndicesRequest and the use of the alias name in the Engine class.

File: x-pack/plugin/ent-search/qa/rest/roles.yml (hunk 9 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/qa/rest/roles.yml:6]: I see, that's because we use the wrong `indices()` in the get engine request. It should be the engine alias name and not the system index name as I outlined in my comment below.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    private GetEngineAction() {
+        super(NAME, GetEngineAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest.Replaceable {
+
+        public static final IndicesOptions DEFAULT_INDICES_OPTIONS = IndicesOptions.strictExpandOpen();
+
+        private String[] names = new String[] { ENGINE_CONCRETE_INDEX_NAME };
```
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:41]: Good catch, thank you!

File: x-pack/plugin/ent-search/qa/rest/roles.yml (hunk 9 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/qa/rest/roles.yml:6]: It looks like it's still required in order for the tests to work. 

Presumably, this is because the system index is protected and the alias is set on the protected system index when it's created? I wasn't able to see concretely where in the code this would be enforced though. Do you have other ideas?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java [hunk 11]
```diff
@@ -77,18 +94,25 @@ public void writeTo(StreamOutput out) throws IOException {
...
     public static final ParseField ANALYTICS_COLLECTION_NAME_FIELD = new ParseField("analytics_collection_name");
+    public static final ParseField UPDATED_AT_MILLIS_FIELD = new ParseField("updated_at_millis");
+    public static final ParseField ENGINE_ALIAS_NAME_FIELD = new ParseField("engine_alias");
     public static final ParseField BINARY_CONTENT_FIELD = new ParseField("binary_content");
 
     static {
         PARSER.declareStringArray(constructorArg(), INDICES_FIELD);
         PARSER.declareStringOrNull(optionalConstructorArg(), ANALYTICS_COLLECTION_NAME_FIELD);
+        PARSER.declareLong(optionalConstructorArg(), UPDATED_AT_MILLIS_FIELD);
+        PARSER.declareString(optionalConstructorArg(), ENGINE_ALIAS_NAME_FIELD);
```
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:115]: I can't think of a way to return the alias name in the response without this?

File: x-pack/plugin/ent-search/qa/rest/roles.yml (hunk 9 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/qa/rest/roles.yml:6]: This is because this PR uses `ENGINE_ALIAS_NAME` for the `indices` in the get request. It should use `Engine.getEngineAliasName(engineName)` instead. We don't want to use the system index alias name, this index is a protected system index that only the engine origin can access. What we expect here is that the security checks the permission on the alias name of the newly created index.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:115]: I don't think we need to return the Engine alias as part of the response, as Jim is saying - is there any other place where you need the alias, as we can use the `getEngineAliasName` method for retrieving that? (Should that be an Engine class method?)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java [hunk 12]
```diff
@@ -92,7 +92,7 @@ public String[] indices() {
 
         @Override
         public IndicesRequest indices(String... indices) {
-            return new Request(new Engine(engine.name(), indices, engine.analyticsCollectionName()));
+            return new Request(new Engine(engine.name(), indices, engine.analyticsCollectionName(), System.currentTimeMillis()));
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:95]: Do we need the update time to be part of the request? It shouldn't be something that can be specified on the request, so I think it should be internal to the service...

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java [hunk 13]
```diff
@@ -62,7 +63,7 @@ protected Collection<Class<? extends Plugin>> getPlugins() {
     }
 
     public void testCreateEngine() throws Exception {
-        final Engine engine = new Engine("my_engine", new String[] { "index_1" }, null);
+        final Engine engine = new Engine("my_engine", new String[] { "index_1" }, null, UPDATED_AT);
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java:66]: We could check that the updated at result is greater than the time the test started, for example, so we don't have to provide updated at as part of the request

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+        @Override
+        public IndicesRequest indices(String... indices) {
+            this.names = indices;
+            return this;
+        }
+
+        @Override
+        public String[] indices() {
+            return names;
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:97]: As Jim was mentioning - this should return the Engine alias name. We probably want that method to be part of the Engine class so it's easier to access from other parts of the code.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(engineId);
+        }
+
+        @Override
+        public IndicesRequest indices(String... indices) {
+            this.names = indices;
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:91]: @jimczi , I guess this method won't be used since we won't be using wildcards or index resolution from the indices() method, and that method will always return the Engine alias.

Could you please explain what can trigger this method to be invoked? What should we do in case it is, update the Engine alias name? 🤔

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:115]: Yes, per [this discussion](https://github.com/elastic/enterprise-search-team/issues/3921#issuecomment-1441905253) we want the alias to be returned to Kibana when searching.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 12 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:95]: @jimczi requested that I have a single constructor on Engine, so I removed the convenience constructor that would not specify the `updated_at` time.

File: x-pack/plugin/ent-search/qa/rest/roles.yml (hunk 9 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/qa/rest/roles.yml:6]: 🤦‍♀️ Thanks, fixing.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java (hunk 14 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:98]: Fixed.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 12 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java:95]: Got it! So Engine will need the `updated_at` in the constructor. But, should the `Request` have the `updated_at` information, given that it is never provided as part of the request? I think we could remove that from the request, and specify at Engine creation. WDYT?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java (hunk 15 above)
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:92]: > Could you please explain what can trigger this method to be invoked?
It should happen only if there are index patterns in the list of indices. Since we disallow index patterns, I like @kderusso 's idea to throw an `IllegalStateException` in this method if that's possible.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 12 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java]: OK, in order to do this I took `updated_at` out of the Engine constructor and replaced it with a setter, let me know what you think.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java [hunk 16]
```diff
@@ -188,7 +186,7 @@ public void getEngine(String resourceName, ActionListener<Engine> listener) {
     }
 
     private static String getEngineAliasName(Engine engine) {
-        return ENGINE_ALIAS_PREFIX + engine.name();
+        return Engine.getEngineAliasName(engine.name());
```
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java]: Good catch, fixed.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java (hunk 15 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:92]: Does it make sense for this class to implement just `IndicesRequest` instead of `IndicesRequest.Replaceable` in that case?

**Review by carlosdelest — APPROVED** [review]

**Review by jimczi — APPROVED** [review]
This LGTM, let's have a follow up PR for the createdAt and updatedAt mutability.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java (hunk 12 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java]: As discussed offline, we'll leave future updates to `updated_at` to a future PR.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java (hunk 15 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/GetEngineAction.java:92]: I'm going to punt on this change for now, simply because making `names` final and throwing made the tests fail. We can revisit it another time though.
